idx/cassandra/cassandra.go

```diff
@@ -546,55 +560,139 @@ func (c *CasIdx) ArchiveDefs(defs []schema.MetricDefinition) (int, error) {
-		pre := time.Now()
-		success = false
-		attempts = 0
-
-		for !success {
-			if err := c.Session.Query(
-				qry,
-				req.def.Id.String(),
-				req.def.OrgId,
-				req.def.Partition,
-				req.def.Name,
-				req.def.Interval,
-				req.def.Unit,
-				req.def.Mtype,
-				req.def.Tags,
-				req.def.LastUpdate).Exec(); err != nil {
-
-				statQueryInsertFail.Inc()
-				errmetrics.Inc(err)
-				if (attempts % 20) == 0 {
-					log.Warnf("cassandra-idx: Failed to write def to cassandra. it will be retried. %s. the value was: %+v", err, *req.def)
+	for {
+		select {
+		case req = <-c.writeQueue:
+			if err != nil {
+				log.Errorf("cassandra-idx: failed to marshal metricDef: %s. value was: %+v", err, *req.def)
+				continue
+			}
+			statQueryInsertWaitDuration.Value(time.Since(req.recvTime))
+			pre := time.Now()
+			success = false
+			attempts = 0
+
+			for !success {
+				select {
+				case <-c.shutdown:
+					log.Info("cassandra-idx: received shutdown, exiting processWriteQueue")
+					return
+				default:
+					c.sessionLock.RLock()
+					if err := c.Session.Query(
+						qry,
+						req.def.Id.String(),
+						req.def.OrgId,
+						req.def.Partition,
+						req.def.Name,
+						req.def.Interval,
+						req.def.Unit,
+						req.def.Mtype,
+						req.def.Tags,
+						req.def.LastUpdate).Exec(); err != nil {
+
+						statQueryInsertFail.Inc()
+						errmetrics.Inc(err)
+						if (attempts % 20) == 0 {
+							log.Warnf("cassandra-idx: Failed to write def to cassandra. it will be retried. %s. the value was: %+v", err, *req.def)
+						}
+						sleepTime := 100 * attempts
+						if sleepTime > 2000 {
+							sleepTime = 2000
+						}
+						time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+						attempts++
+					} else {
+						success = true
+						statQueryInsertExecDuration.Value(time.Since(pre))
+						statQueryInsertOk.Inc()
+						log.Debugf("cassandra-idx: metricDef %s saved to cassandra", req.def.Id)
+					}
+					c.sessionLock.RUnlock()
 				}
-				sleepTime := 100 * attempts
-				if sleepTime > 2000 {
-					sleepTime = 2000
+			}
+		case <-c.shutdown:
+			log.Info("cassandra-idx: received shutdown, exiting processWriteQueue")
+			return
+		}
+	}
+}
+
+func (c *CasIdx) deadConnectionCheck() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(time.Second * 5)
+	attempts := 0
+	totaltime := 0
+	var err error
+
+OUTER:
+	for {
+		// connection to cassandra has been down for at least 30 seconds
+		if attempts >= 6 {
+			c.sessionLock.Lock()
+			for {
+				select {
+				case <-c.shutdown:
+					log.Info("cassandra-idx: received shutdown, exiting deadConnectionCheck")
+					if c.Session != nil && !c.Session.Closed() {
+						c.Session.Close()
+					}
+					// make sure we unlock the sessionLock before returning
+					c.sessionLock.Unlock()
+					return
+				default:
+					log.Errorf("cassandra-idx: creating new session to cassandra using hosts: %v", c.Config.Hosts)
+					if c.Session != nil && !c.Session.Closed() {
+						c.Session.Close()
```

Comment:
closing the old session can be done after the new session has been initialized, assigned to `c.Session`, and the `sessionLock` released. i think it's better to do it this way because closing the session involves acquiring internal gocql locks and pushing objects into internal channels, so it has the potential to block.